### PR TITLE
feat!: dot operators and js opt chaining

### LINKED ISSUE #6201: Semgrep does not support TypeScript  `?` in pattern syntax
**Describe the bug**
In typescript you can write things like this:

`req.user?.accountId` 

Patterns do not accept that as valid and fall over with a parsing error, customers have had pain writing some typescript rules because of this.

**Expected behavior**
support ? as pattern syntax  see the link here https://semgrep.dev/s/1XOk

**Priority**
How important is this to you?
- [x] P2: annoying but not blocking me

### PR DESCRIPTION
**Pull request by brandonspark**
Closes #6201
Closes PA-1943

## What:
This PR adds in a new type to `AST_generic`, called `dot_operator`. This is used to differentiate different kinds of dot accesses.

## Why:
In Javascript, you can do a normal dot access of `x.y`, but also "opt chaining", which looks like `x?.y`. This is essentially just `Option.bind`, but it's a special piece of syntax in the language.

Right now, with just vanilla `DotAccesses`, we have no choice but to match them the same, or to embed the `x` or `y` into weird new constructs. Notably, we can't do this the same way we do it in Rust, where `x?.y` is understood to be the result of `(x?).y`, or the result of doing a dot access on the result of `x?`, because Javascript only has the chaining, and not the standalone postfix `?` operator.

Additionally, adding this new variant instead of offloading this change into the `x` or `y` preserves our current semantic logic as much as possible, because we want to maintain the simple case where the `DotAccess` is taken of something which is a simple identifier.

## How:
Added a new `dot_operator` type, which contains two variants, `Dot` and `QuestDot`. This may be modified in the future to allow more kinds of dot accesses, such as for Kotlin, which has some use cases for this.

PR checklist:

- [X] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [X] Tests included or PR comment includes a reproducible test plan
- [X] Documentation is up-to-date
- [X] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [X] Change has no security implications (otherwise, ping security team)

If you're unsure on any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

**Comment by IagoAbal** [comment]
@brandonspark the CI failures should go away if you rebase against `develop`.

**Comment by IagoAbal** [comment]
@brandonspark remember to add `Closes #NNNN` to the PR comment so that 1) we keep traceability and 2) the issue is closed.

File: changelog.d/pa-1943.fixed [hunk 1]
```diff
@@ -0,0 +1 @@
+JS/TS: Allowed parsing of patterns using the optional chaining "?." syntax.
```
Comment (IagoAbal) [inline changelog.d/pa-1943.fixed]: This should be `gh-6201.fixed`, we should always use the GitHub issue if it exists. Users don't have access to our Linear board.

**Comment by IagoAbal** [comment]
> Notably, we can't do this the same way we do it in Rust, where x?.y is understood to be the result of (x?).y, or the result of doing a dot access on the result of x?, because Javascript only has the chaining, and not the standalone postfix ? operator.

Does it cause any specific issue to translate `x?.y` as `(x?).y` in JS? JS may not have `x?` but our Generic representation does have it (I'm assuming), and _IF_ the semantic meaning is the same then I don't see the problem.

**Comment by aryx** [comment]
Yes, is it possible to unsugar x?.y  in something that would not require to change our DotAccess?
Or we have to change DotAccess?

**Comment by brandonspark** [comment]
Sorry, I requested review again by accident.

File: semgrep-core/src/parsing/tree_sitter/Parse_typescript_tree_sitter.ml [hunk 2]
```diff
@@ -1097,11 +1098,10 @@ and member_expression (env : env) ((v1, v2, v3) : CST.member_expression) : expr
     =
   let expr = expr_or_prim_expr env v1 in
   (* TODO: distinguish optional chaining "?." from a simple access "." *)
```
Comment (nmote) [inline semgrep-core/src/parsing/tree_sitter/Parse_typescript_tree_sitter.ml]: Remove this todo

File: semgrep-core/src/parsing/pfff/js_to_generic.ml [hunk 3]
```diff
@@ -271,7 +271,17 @@ and expr (x : expr) =
...
   | ObjAccess (v1, t, v2) -> (
-      let v1 = expr v1 in
+      let e = expr v1 in
+      let t, is_quest =
+        match t with
+        | Dot, tok -> (info tok, false)
+        | QuestDot, tok -> (info tok, true)
+      in
+      let v1 =
+        if is_quest then
```
Comment (aryx) [inline semgrep-core/src/parsing/pfff/js_to_generic.ml]: you could move all this code in the match above and avoid having this intermediate is_quest.
let v1 = match t with | Dot _ -> e | QuestDot, tok -> G.Call ....

**Review by aryx — APPROVED** [review]
Can you add a comment in AST_generic.ml next to ObjAccess
where you discuss that x?.y is translated in Call Elvis ...
and why we chose this design over adding a new dot_operator type.

**Pull request merged**